File: core-render-android/src/main/java/com/tencent/kuikly/core/render/android/performace/KRPerformanceData.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.core.render.android.performace

import com.tencent.kuikly.core.render.android.performace.frame.KRFrameData
import com.tencent.kuikly.core.render.android.performace.launch.KRLaunchData
import com.tencent.kuikly.core.render.android.performace.memory.KRMemoryData
import org.json.JSONObject

/**
 * 性能数据
 */
data class KRPerformanceData(
    val pageName: String,
    val spentTime: Long,
    val isColdLaunch: Boolean,
    val isPageColdLaunch: Boolean,
    val executeMode: Int,
    val launchData: KRLaunchData?,
    val frameData: KRFrameData?,
    val memoryData: KRMemoryData?
) {
    companion object {
        private const val KEY_MODE = "mode"
        private const val KEY_PAGE_EXIST_TIME = "pageExistTime"
        private const val KEY_IS_FIRST_PAGE_PROCESS = "isFirstLaunchOfProcess"
        private const val KEY_IS_FIRST_PAGE_LAUNCH = "isFirstLaunchOfPage"
        private const val KEY_MAIN_FPS = "mainFPS"
        private const val KEY_KOTLIN_FPS = "kotlinFPS"
        private const val KEY_MEMORY = "memory"
        private const val KEY_PAGE_LOAD_TIME = "pageLoadTime"
    }

    fun toJSONObject(): JSONObject {
        return JSONObject().apply {
            put(KEY_MODE, executeMode)
            put(KEY_PAGE_EXIST_TIME, spentTime)
            put(KEY_IS_FIRST_PAGE_PROCESS, isColdLaunch)
            put(KEY_IS_FIRST_PAGE_LAUNCH, isPageColdLaunch)
            put(KEY_MAIN_FPS, frameData?.getFps() ?: 0)
            put(KEY_KOTLIN_FPS, frameData?.getKuiklyFps() ?: 0)
            launchData?.let {
                put(KEY_PAGE_LOAD_TIME, it.toJSONObject())
            }
            memoryData?.let {
                put(KEY_MEMORY, it.toJSONObject())
            }
        }
    }

    override fun toString() = "[KRLaunchMeta] " +
            "pageName: $pageName, " +
            "spentTime: $spentTime, " +
            "isColdLaunch: $isColdLaunch, " +
            "executeMode: $executeMode"

}

File: core-render-ohos/src/main/cpp/libohos_render/performance/KRPerformanceData.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "KRPerformanceData.h"

#include "thirdparty/cJSON/cJSON.h"

constexpr char kKeyMode[] = "mode";
constexpr char kKeyPageExistTime[] = "pageExistTime";
constexpr char kKeyIsFirstPageProcess[] = "isFirstLaunchOfProcess";
constexpr char kKeyIsFirstPageLaunch[] = "isFirstLaunchOfPage";
constexpr char kKeyMainFPS[] = "mainFPS";
constexpr char kKeyKotlinFPS[] = "kotlinFPS";
constexpr char kKeyMemory[] = "memory";
constexpr char kKeyPageLoadTime[] = "pageLoadTime";

KRPerformanceData::KRPerformanceData(std::string page_name, int excute_mode, int spent_time, bool is_cold_launch,
                                     bool is_page_cold_launch, std::string launch_data)
    : page_name_(page_name), excute_mode_(excute_mode), spent_time_(spent_time), is_cold_launch_(is_cold_launch),
      is_page_cold_launch_(is_page_cold_launch), launch_data_(launch_data) {}

std::string KRPerformanceData::ToJsonString() {
    cJSON *performance_data = cJSON_CreateObject();
    cJSON_AddNumberToObject(performance_data, kKeyMode, excute_mode_);
    cJSON_AddNumberToObject(performance_data, kKeyPageExistTime, spent_time_);
    cJSON_AddBoolToObject(performance_data, kKeyIsFirstPageProcess, is_cold_launch_);
    cJSON_AddBoolToObject(performance_data, kKeyIsFirstPageLaunch, is_page_cold_launch_);
    cJSON_AddStringToObject(performance_data, kKeyPageLoadTime, launch_data_.c_str());
    std::string result = cJSON_Print(performance_data);
    cJSON_Delete(performance_data);
    return result;
}